src/imperative/imperative_utils.cc

```diff
@@ -166,6 +167,16 @@ void RunGraph(
     if (callback) {
         mxnet::common::ExecuteMonOutputCallback(idx, arrays, i, callback);
     }
+    // set the autograd_entry_ in marked nleafs
+    if (nleafs.size()) {
+      auto it = node.source->attrs.dict.find("mark_id");
+      if (it != node.source->attrs.dict.end()) {
+        int mark_id = std::stoi(it->second);
+        CHECK_LT(mark_id, nleafs.size())
+          << "Mark_id exceeds the nonleaf list size.";
+        nleafs[mark_id]->copy_autograd_entry_(ndoutputs[0]);
```

Comment:
Have you verified this is correct if a marked variable has multiple outputs? I see your test case below only considers variables with a single output:

```
        def forward(self, a, b):
            out1 = a*b
            out2 = out1 * a
            self.mark_vars(out1)
            return out2
```

Reply:
I have added the unit test which considers multiple outputs. This is in the commit 'multiple_outputs', also shown below. It shows that it works as expected. 
```
        def forward(self, a, b, c):
            out1 = self.intermediate(('out1_0', 'out1_1'), ((a+b)*c, a*b), grad_req='write')
            out2 = self.intermediate('out2', out1[1] * a)
            return out2
```

In the above commented part, in line 177, indeed the use of `ndoutputs[0]` assumes that in the forward graph each computation node has only one output.  This is consistent with line 152. Based on the test above, for now it seems that this assumption works for the multiple output case: `((a+b)*c, a*b)`.